src/main/java/io/aiven/kafka/connect/gcs/TopicPartitionRecordGrouper.java

```diff
@@ -0,0 +1,131 @@
+ * and {@code start_offset} variables declared.
+ *
+ * <p>The class supports limited and unlimited number of records in files.
+ */
+final class TopicPartitionRecordGrouper implements RecordGrouper {
+    private static final List<String> EXPECTED_VARIABLE_LIST = new ArrayList<>();
+    static {
+        EXPECTED_VARIABLE_LIST.add(FilenameTemplateVariable.TOPIC.name);
+        EXPECTED_VARIABLE_LIST.add(FilenameTemplateVariable.PARTITION.name);
+        EXPECTED_VARIABLE_LIST.add(FilenameTemplateVariable.START_OFFSET.name);
+    }
+    private static final Set<String> EXPECTED_VARIABLE_SET = new HashSet<>(EXPECTED_VARIABLE_LIST);
+
+    private final Template filenameTemplate;
+    private final Integer maxRecordsPerFile;
+
+    private final Map<TopicPartition, SinkRecord> currentFirstRecords = new HashMap<>();
+    private final Map<String, List<SinkRecord>> fileBuffers = new HashMap<>();
+
+    /**
+     * A constructor.
+     *
+     * @param filenameTemplate the filename template.
+     * @param maxRecordsPerFile the maximum number of records per file ({@code null} for unlimited).
+     */
+    public TopicPartitionRecordGrouper(final Template filenameTemplate, final Integer maxRecordsPerFile) {
+        Objects.requireNonNull(filenameTemplate);
+
+        if (!EXPECTED_VARIABLE_SET.equals(new HashSet<>(filenameTemplate.getVariables()))) {
+            throw new IllegalArgumentException(
+                    "filenameTemplate must have set of variables {"
+                            + String.join(",", EXPECTED_VARIABLE_LIST)
+                            + "}, but {"
+                            + String.join(",", filenameTemplate.getVariables())
+                            + "} was given"
+                    );
+        }
+
+        this.filenameTemplate = filenameTemplate;
+        this.maxRecordsPerFile = maxRecordsPerFile;
+    }
+
+    @Override
+    public void put(final SinkRecord record) {
+        Objects.requireNonNull(record);
+
+        final TopicPartition tp = new TopicPartition(record.topic(), record.kafkaPartition());
+        if (!currentFirstRecords.containsKey(tp)) {
+            currentFirstRecords.put(tp, record);
+        }
+        final SinkRecord currentFirstRecord = currentFirstRecords.get(tp);
+
+        final String filename = filenameTemplate.instance()
+                .bindVariable(FilenameTemplateVariable.TOPIC.name, tp::topic)
+                .bindVariable(FilenameTemplateVariable.PARTITION.name, () -> Integer.toString(tp.partition()))
+                .bindVariable(FilenameTemplateVariable.START_OFFSET.name, () -> Long.toString(currentFirstRecord.kafkaOffset()))
+                .render();
+
+        if (!fileBuffers.containsKey(filename)) {
+            fileBuffers.put(filename, new ArrayList<>());
+        }
+
+        if (shouldCreateNewFile(filename)) {
+            currentFirstRecords.put(tp, record);
+            put(record);
```

Comment:
This logic is a bit difficult to understand and looks like it could maybe end up causing problems at some point. At least if new record could ever end up mapping to same file name as earlier record this would result in never ending busy loop.

Reply:
This won't happen with this current implementation, but you're right this recursive implementation is a bit confusing. I changed it.